Add a function for requesting canvas transfer to a thread.

OleksandrChekhovskyi opened the pull request:
At this point, this is an RFC - code might need improvements.

This is to enable use of OffscreenCanvas with pthreads. Calling `emscripten_transfer_canvas()` will make the next created pthread receive the canvas from the main browser thread, and this new thread will be able to issue GL commands directly.

Canvas has to be transferred after the worker is created, but before it started running C++ code (because then it can no longer receive regular JS worker messages), and so the transfer is done in _spawn_thread function.

kripken commented:
This is very cool :)

What's the status of off-screen canvases in browsers - where can this be tested?

OleksandrChekhovskyi commented:
Updated pull request with additional function `emscripten_webgl_commit_context`, for invoking `gl.commit` without EM_ASM.

Simple example program using these functions: https://github.com/OleksandrChekhovskyi/emscripten-offscreen-canvas-test

Currently works in release Firefox with `gfx.offscreencanvas.enabled` and `javascript.options.shared_memory` flags set to true.

Chrome implementation can be tracked here: https://bugs.chromium.org/p/chromium/issues/list?q=label:OffScreenCanvas

OleksandrChekhovskyi commented:
Updated pull request once again, to avoid using addEventListener on OffscreenCanvas (which according to the spec is not an EventTarget). In Firefox right now it is, but in Chrome's work-in-progress implementation it's not.

kripken commented on src/library_gl.js:
```diff
@@ -417,7 +417,8 @@ var LibraryGL = {
       }
       try {
-        canvas.addEventListener('webglcontextcreationerror', onContextCreationError, false);
+        if (canvas.addEventListener)
```
(style nit, need `{}` for ifs etc. that span more than one line)

kripken commented on system/include/emscripten/html5.h:
```diff
@@ -412,6 +412,7 @@ extern void emscripten_webgl_init_context_attributes(EmscriptenWebGLContextAttri
 extern EMSCRIPTEN_WEBGL_CONTEXT_HANDLE emscripten_webgl_create_context(const char *target, const EmscriptenWebGLContextAttributes *attributes);
 
 extern EMSCRIPTEN_RESULT emscripten_webgl_make_context_current(EMSCRIPTEN_WEBGL_CONTEXT_HANDLE context);
+extern EMSCRIPTEN_RESULT emscripten_webgl_commit_context(EMSCRIPTEN_WEBGL_CONTEXT_HANDLE context);
```
will need docs under `site/`.

kripken commented:
We'll also want a test for this, but I guess it's not possible yet?

Overall this looks great to me, nice work!

kripken commented on src/library_pthread.js:
```diff
@@ -404,6 +413,10 @@ var LibraryPThread = {
   },
 
+  emscripten_transfer_canvas: function() {
+    Module['canvasTransferRequested'] = true;
```
is this module property meant to be settable directly as well? if so then should be documented on `site/source/docs/api_reference/module.rst`.

OleksandrChekhovskyi replied:
Not sure about this property, really. I just could not think of a better way to make this work. Alternative was to create a emscripten-specific version of `pthread_create` taking additional options. That seemed too obtuse, but maybe you have a better idea?

kripken replied:
I like how you do it, actually. Seems more convenient to set it on Module in some cases.

OleksandrChekhovskyi replied:
Just to consider all possibilities, if we were to make this a non-public thing, how would we do it then? Also does it actually benefit from being public? I could see a use case where it could be set in shell's html file directly for a trivial case, but eh.

kripken replied:
Not sure what the best way otherwise would be. Maybe @juj has some thoughts.

OleksandrChekhovskyi commented:
Updated pull request to address feedback so far. Added missing documentation, renamed Module property to something more appropriate for public use, `transferCanvas` instead of `canvasTransferRequested`, to confer its imperative nature.

No test so far. Might be tricky to test this reliably.

kripken commented:
Yeah, before consistent browser support it'll be hard to test I guess.

kripken commented:
Overall looks good to me, but @juj knows some of this code better, leaving for him.

juj commented:
Thanks for working on this and for the patience!

Exclusively transferring the control of a canvas to a worker sounds exactly what we want, but the API proposed here to do so feels a bit off. I'd like to propose an alternative API:

  - invariant: At any given time, exactly one thread (main thread or pthread) can have control over a canvas. A page may have multiple canvases, in which case the different threads can control different canvases.
  - if a pthread quits, if it has ownership of any canvases, the ownership of those canvases is transferred back to its parent pthread (or main thread).
  - when a pthread is created with `pthread_create`, instead of using the proposed boolean attribute `Module['transferCanvas'] = true;` to signal WebGL context transferring on hardcoded `Module.canvas` object, let's add a field (e.g. "const char *transferedCanvases;") to `pthread_attr_t` that specifies the DOM IDs of which canvas(es) to transfer. Also, let's have an API function `emscripten_pthread_attr_transfer_canvas(&attr, "#canvasID1, #canvasID2, #canvasID3");` that user can call on a `pthread_attr_t` object before calling `pthread_create(attr)` to specify this list of transfered canvases.
  - a requirement for a thread to be able to transfer a canvas is that that thread must have ownership of that canvas, and the thread shall not have any current WebGL contexts created on that canvas (via call to `emscripten_webgl_create_context()`). If it has, it must `emscripten_webgl_destroy_context()` those contexts first. That is, having any WebGL contexts in existence on a canvas on a given thread "pins" that canvas to that thread and makes it untransferable until the contexts are destroyed.
  - attempting to create a WebGL context on a canvas in a thread that does not have ownership of that canvas shall gracefully fail with an error `EMSCRIPTEN_RESULT_UNKNOWN_TARGET`.

The above scheme would naturally lend to the possibly more complex scenario, where one might create a pthread, which creates another pthread, to which the canvases are transferred in a chain.

The function `emscripten_webgl_commit_context()` looks good, although perhaps the name `emscripten_webgl_commit_frame()` might sound better? Also, let's omit the context handle parameter there, but instead have it operate on the currently active context (that is chosen via `emscripten_webgl_make_context_current()`), since it is essentially an exclusive operation on a context.

A future feature that will be good to have is to support `emscripten_webgl_commit_frame()` even on the main thread, to enable explicit swap control semantics, as opposed to the "implicit-swap-on-return-from-requestAnimationFrame" behavior that currently exists, but that can be separated from this PR. This will be useful especially in conjunction with `--proxy-to-worker` feature.

Another future feature is that by proxying we can actually share the WebGL contexts without needing to manage exclusive access per pthread. But since that will have a different kind of performance landscape, it will be an optional feature, though good to make these two modes compatible from the start.

How does this proposal sound? If you agree, but don't have time to work on that, I can also take this on, whichever you'd prefer? Thanks again for pushing on this feature!

OleksandrChekhovskyi commented:
The proposal sounds much more complicated than originally suggested. Here are some points to consider:

> if a pthread quits, if it has ownership of any canvases, the ownership of those canvases is transferred back to its parent pthread (or main thread).

There is currently no way to transfer control back from OffscreenCanvas to HTMLCanvasElement (will need changes to the spec, and implementation in the browsers).

> Also, let's have an API function emscripten_pthread_attr_transfer_canvas(&attr, "#canvasID1, #canvasID2, #canvasID3");

Element IDs only make sense for HTMLCanvasElements, but not OffscreenCanvas. It will need a way to identify OffscreenCanvas in a pthread, for either another transfer, or for creating a WebGL context.

> The above scheme would naturally lend to the possibly more complex scenario, where one might create a pthread, which creates another pthread, to which the canvases are transferred in a chain.

Since workers can't pass JS-objects to each other directly, canvas will have to be first transferred to the browser thread, and then to another pthread. Which adds more complexity for what seems to be a fringe case.

To me it seems a bit over-engineered. Maybe it's easier to add complexity as it's needed?

juj commented:
> There is currently no way to transfer control back from OffscreenCanvas to HTMLCanvasElement (will need changes to the spec, and implementation in the browsers).

This can be treated as an implementation-specific limitation, not a big issue. The major use case is if a main thread might create a rendering thread, then shut down that thread, and later create another rendering thread. The main thread might not be rendering at all, so going back to HTMLCanvasElement might not be necessary in all cases. Though that is a good point to raise to the spec - does transferring to OffscreenCanvas permanently "taint" the canvas and one would need to delete and recreate the DOM node to revert that?

> Element IDs only make sense for HTMLCanvasElements, but not OffscreenCanvas. It will need a way to identify OffscreenCanvas in a pthread, for either another transfer, or for creating a WebGL context.

We can reuse the element IDs, i.e. if '#mycanvas' was transferred to offscreen, we can keep calling that OffscreenCanvas with the ID '#mycanvas'.

> Since workers can't pass JS-objects to each other directly, canvas will have to be first transferred to the browser thread, and then to another pthread. Which adds more complexity for what seems to be a fringe case.

We do already have this messaging chain implemented, since the same kind of chaining occurs when a pthread creates a pthread, so we should be able to bag the canvas in the existing messages that are passed. Passing the context from main thread -> pthread -> another pthread will happen, since there are Emscripten users who manually make the main thread a dummy one, since it has a restriction of not being able to run one's own blocking main loop, so they create a pthread in start of `main()` to act as the main thread. If they wanted to add a rendering thread into the mix, they'd need a way to pass that through.

> To me it seems a bit over-engineered. Maybe it's easier to add complexity as it's needed?

Do you agree with the model from the application developer's perspective? If so, I can take on the task of implementing this.

OleksandrChekhovskyi commented:
Just for the record, this aspect:

> Passing the context from main thread -> pthread -> another pthread will happen, since there are Emscripten users who manually make the main thread a dummy one, since it has a restriction of not being able to run one's own blocking main loop, so they create a pthread in start of main() to act as the main thread. If they wanted to add a rendering thread into the mix, they'd need a way to pass that through.

was dealt with in the originally suggested design as well. Pthread could request the transfer _from the browser thread_ to another thread. Designated render threads do not actually require a transfer from one pthread to another - simply being able to _request_ that transfer from the browser thread is enough.

Currently OffscreenCanvas support is not mission-critical for me: as it works only in Firefox and has bugs, for now I proxy GL calls to the browser thread instead (custom implementation in native code, as proxying code in Emscripten expects worker to run the event loop). So for now I'm not ready to commit to implementing a much more complex solution - and therefore it makes more sense if you take over.

The second pull request waiting review ("running async callbacks on the main thread") is what allows me to proxy things directly in synchronous native code, which is how I solved the WebGL case.

juj commented:
Fair enough, agree on your comments. Posted an implementation of the `pthread_attr_t`-based canvas transfer machinery in PR #4412. Let me know if there are some specific concerns you have there. Two test cases stress the scenario of main thread -> thread1 -> thread2 chaining, and the main thread -> thread1 -> back to main thread chaining.

You mention that you found OffscreenCanvas buggy in Firefox: are these reported somewhere? I hit two different issues, neither of which were blocking for testing the PR:
   - https://bugzilla.mozilla.org/show_bug.cgi?id=1281909
   - https://bugzilla.mozilla.org/show_bug.cgi?id=1281796

Would be helpful to track the other known issues here as well.

OleksandrChekhovskyi commented:
https://bugzilla.mozilla.org/show_bug.cgi?id=1273508

This is the main issue I've seen. Usage of OffscreenCanvas leaks resources and prevents clean exit of Firefox. That is a terrible UX, so it would prevent any real usage of the feature until it's fixed.

Also compositing performance in full screen seems to be really subpar on OS X (on Windows it's fine).

juj commented:
Thanks for posting the bug report. A bug is a bug, and not ever really part of UX design, so "terrible UX" is not surprising. This is after all an evolving and actively implemented spec, so critical crash bugs are not unusual. Like you mentioned in an earlier comment, proxying the GL calls to the main thread is a more compatible and stable way to get pthread rendering going. I'll take a look at that PR asap once this and a few other PRs have landed.

It sounds like the PR #4412 will adequately solve this, so let me close this PR.

OleksandrChekhovskyi commented:
Yeah, "UX" was indeed a poor choice of a term. The point was that end user running an OffscreenCanvas application will have a bad experience with leaks, so launching a public preview of an app with this would be out of question, until it's fixed.
